src/RA_Dlg_MemBookmark.cpp

```diff
@@ -704,50 +704,74 @@ void Dlg_MemBookmark::ExportJSON()
     {
-        hr = pDlg->SetFileTypes(ARRAYSIZE(c_rgFileTypes), c_rgFileTypes);
-        if (hr == S_OK)
+		
+		if (hr = pDlg->SetFileTypes(c_rgFileTypes.size(), &c_rgFileTypes.front()); SUCCEEDED(hr))
         {
-            char defaultFileName[512];
-            sprintf_s(defaultFileName, 512, "%s-Bookmarks.txt", std::to_string(g_pCurrentGameData->GetGameID()).c_str());
-            hr = pDlg->SetFileName(Widen(defaultFileName).c_str());
-            if (hr == S_OK)
+            std::ostringstream oss;
+            oss.str().reserve(512); // storage on the heap really shouldn't matter but keeping relativelty the same
+            oss << g_pCurrentGameData->GetGameID() << "-Bookmarks.txt";
+
+            auto defaultFileName{ oss.str() };
+            // get in the habit of reseting the stringbuffer inside the
+            // stringstream if you wish to use it more than once (you will get
+            // garabage characters in the end)
+            oss.str(""); 
+
+
+
+			if (hr = pDlg->SetFileName(Widen(defaultFileName).c_str()); SUCCEEDED(hr))
             {
-                PIDLIST_ABSOLUTE pidl;
-                hr = SHParseDisplayName(Widen(defaultDir).c_str(), nullptr, &pidl, SFGAO_FOLDER, 0);
-                if (hr == S_OK)
+                // nullptr is a literal for pointers, just like "s" is for string or "ULL" is for (unsigned long long)
+                // Since every pointer can be converted to std::nullptr_t, every pointer is a NullablePointer
+                // http://en.cppreference.com/w/cpp/concept/NullablePointer
+                PIDLIST_ABSOLUTE pidl{ nullptr };
+        
+				if (hr = SHParseDisplayName(Widen(defaultDir).c_str(), nullptr, &pidl, SFGAO_FOLDER, nullptr); SUCCEEDED(hr))
                 {
-                    IShellItem* pItem = nullptr;
-                    SHCreateShellItem(nullptr, nullptr, pidl, &pItem);
-                    hr = pDlg->SetDefaultFolder(pItem);
-                    if (hr == S_OK)
+					CComPtr<IShellItem> pItem;
+					SHCreateShellItem(nullptr, nullptr, pidl, &pItem );
+					
+					if (hr = pDlg->SetDefaultFolder(pItem); SUCCEEDED(hr))
                     {
                         pDlg->SetDefaultExtension(L"txt");
-                        hr = pDlg->Show(nullptr);
-                        if (hr == S_OK)
+						
+						if (hr = pDlg->Show(nullptr); SUCCEEDED(hr))
                         {
 
-                            hr = pDlg->GetResult(&pItem);
-                            if (hr == S_OK)
-                            {
+                            // Exception got thrown here (didn't before), -sbs
+                            // Expression: p==0; where p is pItem.p
+                            // Solution: reset the owned pointer
+                            auto pItem2 = std::move(pItem); // let's see if that'll do it
+
+							if (hr = pDlg->GetResult(&pItem); SUCCEEDED(hr))
+                            {                              
```

Comment:
It looks like this code could use a reformatting pass. That should be safe now that #39 has been merged.

Reply:
I'll merge RetroAchievements/master with this branch preferring the remote branch as far as whitespace is concerned. That should fix the whitespace issue.